### ISSUE TITLE
Incorrect logic for re_capture (and friends)?

### ISSUE THREAD

**Issue by tmccombs** [Geal/nom#482]
The logic in https://github.com/Geal/nom/blob/master/src/regexp.rs#L295 seems wrong to me. It seems like it should use `v.first()` to get the full match rather than `v.last()` which just gets the last capture group which may or may not actually be at the end of the match.

*Note:* I'd be happy to make a PR, I'm just not sure if I'm missing something.

**Comment by Geal** [comment]
Hello,

`first()` will return the first match, which will probably not go up to the end either. I think the potential issue here is that the code assumes than no other match than the last starts this far, and no other match than the last goes this far. A cleaner way might be to accumulate the `m.end()` values and keep the largest one.

**Comment by tmccombs** [comment]
> Capture group 0 always corresponds to the entire match.
(from https://docs.rs/regex/0.2.5/regex/struct.Regex.html#method.captures)

So the first "capture group" is actually the entire match.

See https://play.rust-lang.org/?gist=69d4830a24288c60b2a721c13bd6629b&version=stable

**Comment by Geal** [comment]
oh, right, sorry, since you linked to re_bytes_matches, I forgot you talked about the ones using captures.
Do you have examples I could add to the unit tests, to make sure the fix works?

**Comment by tmccombs** [comment]
>  since you linked to re_bytes_matches
sorry about that. the line numbers must have changed since I first posted the issue. I'll be honest, it confused me too, when I followed the link.

Unfortunately, I don't remember the use case that inspired me to file this (and I found another way to achieve my goal at the time). However, any regex with characters after the last capture group could be used to test this, for example: `\[([a-z]+)\]`, `([a-z]+)\s*` (tested against a string with trailing whitespace), `(\d+)_[a-z]`, etc.
